MinecraftClient/McClient.cs

```diff
@@ -1084,6 +1085,102 @@ public bool MoveTo(Location location, bool allowUnsafe = false, bool allowDirect
+        /// <param name="radius">Maximum distance of the approach location to the goal</param>
+        /// <param name="pathfindingThreads">How many threads should be started to find valid paths</param>
+        /// <param name="minDistance">Minimum distance of the approach location to your goal - must be smaller than radius</param>
+        /// <param name="allowUnsafe">Allow non-vanilla direct teleport instead of computing path, but may cause invalid moves and/or trigger anti-cheat plugins</param>
+        /// <returns>True if a location has been found and movement is started</returns>
+        public bool ApproachTo(Location goalToApproach, int radius, int pathfindingThreads=10, double minDistance=0, bool allowUnsafe=false) 
+        {
+            // Set border of the search area
+            Location minPoint = new Location(goalToApproach.X - radius, goalToApproach.Y - radius, goalToApproach.Z - radius);
+            Location maxPoint = new Location(goalToApproach.X + radius, goalToApproach.Y + radius, goalToApproach.Z + radius);
+
+            // Create a list of all blocks that should be searched
+            List<int> xRange = Enumerable.Range(Convert.ToInt32(Math.Floor(minPoint.X)), Convert.ToInt32(Math.Floor(maxPoint.X - minPoint.X)) + 1).ToList();
+            List<int> yRange = Enumerable.Range(Convert.ToInt32(Math.Floor(minPoint.Y)), Convert.ToInt32(Math.Floor(maxPoint.Y - minPoint.Y)) + 1).ToList();
+            List<int> zRange = Enumerable.Range(Convert.ToInt32(Math.Floor(minPoint.Z)), Convert.ToInt32(Math.Floor(maxPoint.Z - minPoint.Z)) + 1).ToList();
+
+            List<Location> listOfBlocks = xRange.SelectMany(x => yRange.SelectMany(y => zRange.Select(z => new Location(x, y, z)))).ToList();
+
+            // Get all blocks that are air, have an air block above and a solid block below (space for a player to fit in)
+            // and sort them after their distance to goalToApproach. Take 5 by default to avoid too much calculation power.
+            List<Location> listOfCloseLocations = listOfBlocks.Where(block => GetWorld().GetBlock(new Location(block.X, block.Y + 1, block.Z)).Type == Material.Air &&      // AIR
+                                                                            GetWorld().GetBlock(block).Type == Material.Air &&                                              // AIR
+                                                                            GetWorld().GetBlock(new Location(block.X, block.Y - 1, block.Z)).Type.IsSolid())                // SOLID
+                                                                .OrderBy(closeLocation => goalToApproach.Distance(closeLocation)).ToList();
+
+            // If a minimal distance to the goal was given, remove all locations that are too close.
+            if (minDistance>0) 
+            {
+                listOfCloseLocations = listOfCloseLocations.Where(block => block.Distance(goalToApproach) >= minDistance).ToList();

```

Comment:
Note that Distance does `sqrt()`, which is slow. `radius` being already a cube, maybe you can use `DistanceSquared`. Multiply `minDistance` by itself once: `minDistanceSquared = minDistance * minDistance`, then compare with `DistanceSquared`, distance will be a cube instead of a sphere. (Also I'm not sure why you'd want a minimum distance. Maybe for cactus? But `Movement.IsSafe()` should detect it)